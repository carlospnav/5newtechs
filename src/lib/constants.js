export const fetchPolicyList = {
  fetchPolicy: "network-only"
}
