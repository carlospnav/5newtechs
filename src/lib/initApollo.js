import { ApolloClient } from "apollo-client"
import { HttpLink } from "apollo-link-http"
import { InMemoryCache } from "apollo-cache-inmemory"
import fetch from "isomorphic-fetch"
import gql from "graphql-tag"
let apolloClient = null

if (!process.browser) {
  global.fetch = fetch
}

function create({ state }) {
  const httpLink = new HttpLink({
    ssrMode: !process.browser,
    uri: process.env.GRAPHQL_API_URL
  })

  return new ApolloClient({
    link: httpLink,
    cache: new InMemoryCache().restore(state)
  })
}

export default function initApollo(options) {
  if (!process.browser) {
    return create(options)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = create(options)
  }

  return apolloClient
}
