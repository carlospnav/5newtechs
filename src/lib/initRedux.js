import { createStore, combineReducers, compose } from "redux"
import url from "./reducers/url"

let reduxStore = null

// Get the Redux DevTools extension and fallback to a no-op function
let devtools = f => f
if (process.browser && window.__REDUX_DEVTOOLS_EXTENSION__) {
  devtools = window.__REDUX_DEVTOOLS_EXTENSION__()
}

function create() {
  return createStore(
    combineReducers({
      // Setup reducers
      url
    }),
    compose(devtools)
  )
}

export default function initRedux() {
  if (!process.browser) {
    return create()
  }

  if (!reduxStore) {
    reduxStore = create()
  }

  return reduxStore
}
