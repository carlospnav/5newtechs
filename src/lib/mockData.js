export const films = [
  {
    id: "1",
    episodeID: 1,
    title: "A New Hope",
    releaseDate: "2010"
  },
  {
    id: "2",
    episodeID: 2,
    title: "A First Hope",
    releaseDate: "2011"
  },
  {
    id: "3",
    episodeID: 3,
    title: "A Last Hope",
    releaseDate: "2012"
  }
]

export const film = {
  id: "1",
  episodeID: 1,
  title: "A First Hope Returns Wars",
  release: "2009",
  content: "Damn you, openingCrawl!",
  director: "Michael Bay"
}

export const person = {
  personId:"1",
  name: "Tauntaun",
  homeworld: {
    name: "Hoth"
  },
  gender: "Male",
  birthYear: "1001",
  skinColor: "White",
  hairColor: "White",
  eyeColor: "Black"
}

export const people = [
  {
    personId:"1",
    name: "Tauntaun",
    homeworld: {
      name: "Hoth"
    },
    gender: "Male",
    birthYear: "1001",
    skinColor: "White",
    hairColor: "White",
    eyeColor: "Black"
  },
  {
    personId:"2",
    name: "Stormtrooper's stray blaster shot",
    homeworld: {
      name: "Imperial Blaster"
    },
    gender: "Neutral",
    birthYear: "0000",
    skinColor: "Translucent",
    hairColor: "Translucent",
    eyeColor: "Redish"
  }
]

