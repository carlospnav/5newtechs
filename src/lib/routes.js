const routes = (module.exports = require("next-routes")())

routes.add("people").add("person_detail", "/people/:personId").add("movies", "/movies/:filmId")
