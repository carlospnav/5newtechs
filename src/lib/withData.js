import { Component } from "react"
import { Provider } from "react-redux"
import { getDisplayName } from "recompose"
import { ApolloProvider, getDataFromTree } from "react-apollo"
import initApollo from "./initApollo"
import initRedux from "./initRedux"

function App({ ComposedComponent, redux, apollo, ...props }) {
  return (
    <ApolloProvider client={apollo}>
      <Provider store={redux}>
        <ComposedComponent {...props} />
      </Provider>
    </ApolloProvider>
  )
}

export default ComposedComponent => {
  return class WithData extends Component {
    static displayName = `WithData(${getDisplayName(ComposedComponent)})`


    static getInitialProps(ctx) {
      if (process.browser) {
        return {}
      }

      return new Promise(async resolve => {
        const composedInitialProps = ComposedComponent.getInitialProps
          ? await ComposedComponent.getInitialProps(ctx)
          : {}

        const redux = initRedux()
        const url = { query: ctx.query, pathname: ctx.pathname }
        const apollo = initApollo({})

        try {
          await getDataFromTree(
            <App
              ComposedComponent={ComposedComponent}
              redux={redux}
              apollo={apollo}
              url={url}
              {...composedInitialProps}
            />
          )
        } catch (err) {}

        resolve({ apolloState: apollo.extract() })
      })
    }

    constructor(props) {
      super(props)

      this.redux = initRedux()
      this.apollo = initApollo({ state: props.apolloState })
    }

    render() {
      return (
        <App
          ComposedComponent={ComposedComponent}
          redux={this.redux}
          apollo={this.apollo}
          {...this.props}
        />
      )
    }
  }
}
