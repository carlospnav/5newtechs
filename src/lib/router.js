import { Router as NextRouter } from "lib/routes"

export default function Router({ returnTo }) {
  return {
    goToPeople() {
      return NextRouter.pushRoute("people")
    },
    goBack(orCallThis) {
      if (returnTo) {
        return () => NextRouter.pushRoute(returnTo)
      } else {
        return orCallThis
      }
    }
  }
}
