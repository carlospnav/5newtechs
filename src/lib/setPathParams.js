import { Router } from "lib/routes"
import pickBy from "lodash/pickBy"
import includes from "lodash/includes"
import isEmpty from "lodash/isEmpty"
import split from "lodash/split"
import queryString from "query-string"

export default function setPathParams(params, state) {
  const queryParams = pickBy(params, (value, key) => {
    const previousValue = state.params[key]
    if (!value) {
      return false
    } else if (isEmpty(previousValue)) {
      return true
    } else {
      return includes(state.path, key)
    }
  })
  const urlQuery = queryString.stringify(queryParams)
  const pathWithoutQuery = split(state.path, "?")[0]
  const separator = urlQuery.length > 0 ? "?" : ""
  const newPath = `${pathWithoutQuery}${separator}${urlQuery}`
  if (typeof window !== "undefined") {
    Router.replaceRoute(newPath, params, { shallow: true })
  }
  return newPath
}
