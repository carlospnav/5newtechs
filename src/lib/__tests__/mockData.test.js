import { films, film, people, person } from "lib/mockData"

describe("MockData", () => {

  it("films should match the snapshot", () => {
    expect(films).toMatchSnapshot()
  })

  it("film should match the snapshot", () => {
    expect(film).toMatchSnapshot()
  })

  it("people should match the snapshot", () => {
    expect(people).toMatchSnapshot()
  })

  it("person should match the snapshot", () => {
    expect(person).toMatchSnapshot()
  })
})