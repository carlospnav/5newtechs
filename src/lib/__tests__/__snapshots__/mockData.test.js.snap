// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MockData film should match the snapshot 1`] = `
Object {
  "content": "Damn you, openingCrawl!",
  "director": "Michael Bay",
  "episodeID": 1,
  "id": "1",
  "release": "2009",
  "title": "A First Hope Returns Wars",
}
`;

exports[`MockData films should match the snapshot 1`] = `
Array [
  Object {
    "episodeID": 1,
    "id": "1",
    "releaseDate": "2010",
    "title": "A New Hope",
  },
  Object {
    "episodeID": 2,
    "id": "2",
    "releaseDate": "2011",
    "title": "A First Hope",
  },
  Object {
    "episodeID": 3,
    "id": "3",
    "releaseDate": "2012",
    "title": "A Last Hope",
  },
]
`;

exports[`MockData people should match the snapshot 1`] = `
Array [
  Object {
    "birthYear": "1001",
    "eyeColor": "Black",
    "gender": "Male",
    "hairColor": "White",
    "homeworld": Object {
      "name": "Hoth",
    },
    "name": "Tauntaun",
    "personId": "1",
    "skinColor": "White",
  },
  Object {
    "birthYear": "0000",
    "eyeColor": "Redish",
    "gender": "Neutral",
    "hairColor": "Translucent",
    "homeworld": Object {
      "name": "Imperial Blaster",
    },
    "name": "Stormtrooper's stray blaster shot",
    "personId": "2",
    "skinColor": "Translucent",
  },
]
`;

exports[`MockData person should match the snapshot 1`] = `
Object {
  "birthYear": "1001",
  "eyeColor": "Black",
  "gender": "Male",
  "hairColor": "White",
  "homeworld": Object {
    "name": "Hoth",
  },
  "name": "Tauntaun",
  "personId": "1",
  "skinColor": "White",
}
`;
